eoe/man/man1/date.1
'\"macro stdmacro
.if n .pH g1.date @(#)date	41.13 of 5/26/91
.\" Copyright 1991 UNIX System Laboratories, Inc.
.\" Copyright 1989, 1990 AT&T
.nr X
.if \nX=0 .ds x} date 1 "Essential Utilities" "\&"
.if \nX=1 .ds x} date 1 "Essential Utilities"
.if \nX=2 .ds x} date 1 "" "\&"
.if \nX=3 .ds x} date "" "" "\&"
.TH \*(x}
.SH NAME
\f4date\f1 \- write and set the date and time
.SH SYNOPSIS
\f4date\f1 [
\f4-u\fP
]
[
\f4+\fP
.I format
]
.br
\f4date\f1
[
\f4-a\f1
[
\f4\-\f1
]
\f2sss\f4.\f2fff\f1
]
.br
\f4date\f1
[
\f4\-u\f1
|
\f4\-n\f1
]
[\|[
.IR mmdd \|]\| HHMM
|
.IR mmddHHMM \|[\| cc \|]\| yy 
]
[
.IR .ss
]
.SH DESCRIPTION
If no argument is given,
or if the argument begins with \f4+\f1,
the current date and time are printed.
Otherwise, the current date is set
if the user has appropriate privileges and the system permits 
the setting of the system date and time.
.P
Supplementary code set characters in \f4+\f2format\f1
(see below) are recognized and displayed
according to the locale specified in the \f4LC_CTYPE\fP
environment variable [see \f4LANG\fP on \f4environ\fP(5)].
Month and weekday names are recognized according
to the locale specified in the \f4LC_TIME\f1 environment variable,
as described below.
.SH OPTIONS
.PP
.TP 12
\f4\-a\f1 [ \f4\-\f1 ] \f2sss\f4.\f2fff\f1
Slowly adjust the time by
.IR sss . fff
seconds
.RI ( fff
represents fractions of a second).
This adjustment can be positive or negative.
The system's clock will be sped up or slowed down until it has drifted
by the number of seconds specified.
.TP
\f4\-n\f1
Do not attempt to notify the network time daemon
.IR timed (1m)
of the time change (or more accurately the daemon listening to
the \f4timed\fP UDP port).  This option should be used if you
are not using that daemon.  If not used, the daemon is notified
of the date change, and if it is acting as the master, the date
change will be broadcast to the time daemons on other systems.
.TP
\f4\-u\f1
Display (or set) the date in Greenwich Mean Time
(\s-1GMT\s0\(emuniversal time),
bypassing the normal conversion to
(or from) local time.  
.TP
\f4\ \f1
The notation GMT may change in the future
to the equivalent notation UTC (Coordinated Universal Time).
.TP
.I mm\^
is the month number
.TP
.I dd\^
is the day number in the month
.TP
.I HH\^
is the hour number (24 hour system)
.TP
.I MM\^
is the minute number
.TP
.I ss\^
is the second(s) number
.TP
.I cc\^
is the century minus one
.TP
.I yy\^
is the last 2 digits of the year number
.IP
The month, day, year, and century may be omitted;
the current values are supplied as defaults.
.PP
For example:
.IP "" 18
\f4date\| 10080045\fP
.IP "" 12
sets the date to Oct 8, 12:45 \s-1AM\s+1.
.PP
The current year is the default because no year is supplied.
The system operates in \s-1GMT\s0.
\f4date\fP
takes care of the conversion to and from
local standard and daylight time.
The date can be changed only if the user has appropriate privileges 
and the system permits the setting of the system date and time.
.PP
After successfully setting the date and time,
\f4date\fP
displays the new date according
to the default format.
The \f4date\fP
command uses
\f4TZ\fP
to determine the correct time zone information
[see
\f4environ\fP(5)].
.SH OPERANDS
.TP
\f4+\fP \f2format\fP
If the argument begins with \f4+\fP,
the output of
\f4date\fP
is under the control of the user.
Each Field Descriptor, described below, is preceded by \f4%\fP
and is replaced in the output by
its corresponding value.
A single \f4%\fP is encoded by \f4%%\fP.
All other
characters are copied to the output without change.
The string is always terminated with a newline character.
If the argument contains embedded blanks
it must be quoted (see the
\s-1EXAMPLE\s0
section).
Supplementary code set characters may be used in \f2format\f1.
.PP
As noted, month and weekday names are recognized
according to the locale specified in the \f4LC_TIME\fP
environment variable [see \f4LANG\fP on \f4environ\fP(5)].
The names are taken from a file whose format is specified in
\f4strftime\fP(4).
This file also defines country-specific date and
time formats such as \f4%c\fP,
which specifies the default date format.
The following form is the default for \f4%c\fP:
.sp .5
.RS
.nf
	\f4%a %b %e %T %Z %Y\fP
	e.g., Fri Dec 23 10:10:42 EST 1988
.fi
.RE
.PP
Field Descriptors (must be preceded by a \f4%\fP):
.sp .5
.RS 3
.PD 0
.TP 5
\f4 a\f1
locale's abbreviated weekday name
.TP 5
\f4 A\f1
locale's full weekday name
.TP 5
\f4 b\f1
locale's abbreviated month name
.TP 5
\f4 B\f1
locale's full month name
.TP 5
\f4 c\f1
locale's appropriate date and time representation
.TP 5
\f4 C\f1
century (a year divided by 100 and truncated to an integer) as
a decimal number [00-99]
.TP 5
\f4 d\f1
day of month as a decimal number [01-31]
.TP 5
\f4 D\f1
date as \f4%m/%d/%y\f1
.TP 5
\f4 e\f1
day of month as a decimal number [1-31] (single digits are preceded by a blank)
.TP 5
\f4 h\f1
locale's abbreviated month name (alias for \f4%b\fP)
.TP 5
\f4 H\f1
hour (24-hour clock) as a decimal number [00-23]
.TP 5
\f4 I\f1
hour (12-hour clock) as a decimal number [01-12]
.TP 5
\f4 j\f1
day of year as a decimal number [001-366]
.TP 5
\f4 m\f1
month of year as a decimal number [01-12]
.TP 5
\f4 M\f1
minute as a decimal number [00-59]
.TP 5
\f4 n\f1
insert a newline character
.TP 5
\f4 p\f1
string containing ante-meridian or post-meridian
indicator (by default, AM or PM)
.TP 5
\f4 r\f1
12-hour clock time [01-12] using the AM/PM notation, as \f4%I:%M:%S %p\f1
.TP 5
\f4 R\f1
time as \f4%H:%M\f1
.TP 5
\f4 S\f1
seconds as a decimal number [00-61], allows for leap seconds
.TP 5
\f4 t\f1
insert a tab character
.TP 5
\f4 T\f1
24-hour clock time [00-23], as \f4%H:%M:%S\f1
.TP 5
\f4 u\f1
weekday as a decimal number [1(Monday)-7]
.TP 5
\f4 U\f1
week number of year (Sunday as the first day of the week) 
as a decimal number [00-53]. All days in a new year 
preceding the first Sunday are considered to be in week 0.
.TP 5
\f4 V\f1
week of the year (Monday as the first day of the week) as a decimal number 
[01-53]. If the week containing January 1 has four or more days 
in the new year, then it is considered week 1; 
otherwise, it is week 53 of the previous year, and 
the next week is week 1. ( See the ISO 8601:1988 standard)
.sp 0 
.TP 5
\f4 w\f1
weekday as a decimal number [0(Sunday)-6]
.TP 5
\f4 W\f1
week of year (Monday as the first day of the week) as a decimal number [00-53].
All days in a new year preceding the first Sunday 
are considered to be in week 0.
.TP 5
\f4 x\f1
locale's appropriate date representation
.TP 5
\f4 X\f1
locale's appropriate time representation
.TP 5
\f4 y\f1
year within century (offset from %C) as a decimal number [00-99]
.TP 5
\f4 Y\f1
year as \f2ccyy\f1 (4 digits)
.TP 5
\f4 Z\f1
timezone name, or no characters if no timezone is determinable
.TP 5
\f4 %\f1
insert a percent sign character
.RE
.PD
.PP
Modified Field Descriptors (must be preceded by a \f4%\fP):
.PP
Some field descriptors can be modified by the E and O modifier characters
to indicate a different format or specification as specified in the LC_TIME 
locale description (see the XBD specification). If the corresponding keyword
is not specified or not supported for the current locale, the unmodified
field descriptor value will be used.
.sp .5
.RS 3
.PD 0
.TP 5
\f4 Ec\f1
locale's alternative appropriate date and time representation
.TP 5
\f4 EC\f1
name of the base year (period) in the locale's alternative representation
.TP 5
\f4 Ex\f1
locale's alternative date representation
.TP 5
\f4 EX\f1
locale's alternative time representation
.TP 5
\f4 Ey\f1
offset from %EC (year only) in the locale's alternative representation
.TP 5
\f4 EY \f1
full alternative year representation
.TP 5
\f4 Od\f1
day of the month using the locale's alternative numeric symbols
.TP 5
\f4 Oe\f1
day of the month using the locale's alternative numeric symbols
.TP 5
\f4 OH\f1
hour (24 hour clock) using the locale's alternative numeric symbols
.TP 5
\f4 OI\f1
hour (12 hour clock) using the locale's alternative numeric symbols
.TP 5
\f4 Om\f1
month using the locale's alternative numeric symbols
.TP 5
\f4 OM\f1
minutes using the locale's alternative numeric symbols
.TP 5
\f4 OS\f1
seconds using the locale's alternative numeric symbols
.TP 5
\f4 Ou\f1
weekday as a number in the locale's alternative representation (Monday=1)
.TP 5
\f4 OU\f1
week number of the year (Sunday as first day of the week) 
using the locale's alternative numeric symbols
.TP 5
\f4 OV\f1
week number of the year (Monday as first day of the week, 
rules corresponding to %V) using the locale's alternative numeric symbols
.TP 5
\f4 Ow\f1
weekday as a number in the locale's alternative representation (Sunday=0)
.TP 5
\f4 OW\f1
week number of the year (Monday as first day of the week) 
using the locale's alternative numeric symbols
.TP 5
\f4 Oy\f1
year (offset from %C) in alternative representation

.RE
.PD
.ne 8
.SH EXAMPLE
The command
.IP
\f4date '+DATE: %m/%d/%y%nTIME: %H:%M:%S'\f1
.PP
generates as output:
.sp .5
.nf
	\f4DATE: 08/01/76
	TIME: 14:45:05\f1
.fi
.br
.ne 1.5i
.SH ENVIRONMENT VARIABLES
The following environment variables affect the execution of 
\f4date\fP:
.TP
\f4LANG\fP
provide a default value for the internationalization variables that are unset or null.
If 
.I LANG\^
is unset or null the corresponding value from the implementation-specific
default locale will be used. If any of the internationalization variables 
contains an invalid setting, the utility will behave as if 
none of the variables had been defined.
.TP
\f4LC_ALL\fP
if set to a non-empty string value, override the values of all the 
other internationalization variables.
.TP
\f4LC_CTYPE\fP
determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single - as opposed to multi-byte -
characters in arguments).
.TP
\f4LC_MESSAGES\fP
determine the locale that should be used to affect the format and contents of diagnostic messages written to standard error.
.TP
\f4LC_TIME\fP
determine the format and contents of date and time strings written by
.I date\^
.
.TP
\f4NLSPATH\fP
determine the location of message catalogs for the processing of
.I LC_MESSAGES.\^
.
.TP
\f4TZ\fP
determine the timezone in which the time and date are written, unless the
.B \-u
option is specified. If the 
.I TZ\^
variable is not set and the
.B \-u
is not specified, an unspecified system default timezone is used.
.SH STDOUT
When no formatting operand is specified, the output in the POSIX locale 
is equivalent to specifying:

.sp .5
.nf
        \f4date "+%a %b %e %H:%M:%S %Z %Y" \f1
.fi
.br
.SH FILES
.TP
\f4/usr/lib/locale/\f2locale\f4/LC_MESSAGES/uxcore.abi\f1
language-specific message file [See \f4LANG\fP on \f4environ\f1 (5).]
.SH SEE ALSO
\f4syslog\f1(3),
\f4strftime\f1(4),
\f4timed\f1(1M),
\f4timeslave\f1(1M),
\f4utmp\f1(4),
\f4environ\f1(5)
.SH DIAGNOSTICS
.PD 0
.TP 18
\f5UX:date:ERROR:No permission\f1
You do not have the appropriate privileges
and you try to change the date.
.TP
\f5UX:date:ERROR:bad conversion\f1
The date set is syntactically incorrect.
.PD
.SH NOTES
.P
If you attempt to set the current date to one of the dates 
on which the standard and
alternate time zones change
(for example, the date that daylight time is starting or ending),
and you attempt to set the time to a time in the interval
between the end of standard time and the beginning of the alternate time
(or the end of the alternate time and the beginning of standard time),
the results are unpredictable.
.P
If you are running a network-based time service, such as timed, then
it will override any changes you make with the date command.
.Ee


eoe/cmd/date/date.c
/*	Copyright (c) 1990, 1991 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/*#ident	"@(#)date:date.c	1.24.1.7" */
#ident "$Header: /proj/irix6.5.7m/isms/eoe/cmd/date/RCS/date.c,v 1.19 1997/10/24 23:56:42 jiju Exp $"

/*******************************************************************

		PROPRIETARY NOTICE (Combined)

This source code is unpublished proprietary information
constituting, or derived under license from AT&T's UNIX(r) System V.
In addition, portions of such source code were derived from Berkeley
4.3 BSD under license from the Regents of the University of
California.

		Copyright Notice 

Notice of copyright on this source code product does not indicate 
publication.

	(c) 1986,1987,1988,1989  Sun Microsystems, Inc
	(c) 1983,1984,1985,1986,1987,1988,1989  AT&T.
	          All rights reserved.
********************************************************************/ 

/*
**	date - with format capabilities and international flair
*/

#include        <stdio.h>
#include	<time.h>
#include	<sys/time.h>
#include	<sys/types.h>
#include	<locale.h>
#include	<fcntl.h>
#ifdef sgi
#include	<utmpx.h>	/* includes utmp.h */
#else
#include	<utmp.h>
#endif
#include	<pfmt.h>
#include	<errno.h>
#include	<string.h>


#define	year_size(A)	(((A) % 4) ? 365 : 366)

static 	char	buf[1024];
static	time_t	clock_val;
static  short	month_size[12] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
static  struct  utmp	wtmp[2] = { {"", "", OTIME_MSG, 0, OLD_TIME, 0, 0, 0}, 
			    {"", "", NTIME_MSG, 0, NEW_TIME, 0, 0, 0} };
#ifdef sgi
static	struct	utmpx	wtmpx[2];
extern	void	getutmpx(const struct utmp *, struct utmpx *);
#endif

static  int uflag = 0;
static  int nflag = 0;                  /* !=0 to not set the network time */
static  int vflag = 0;			/* variable format flag */

/*
 * Procedure:     main
 */
main(argc, argv)
int	argc;
char	**argv;
{
	register struct tm *tp;
	struct timeval tv;
	char *fmt = NULL;
	char *locp;
	int retval;		/* scratch variable */
	
	locp = setlocale(LC_ALL, "");
	if (locp == (char *)0) {
		(void)setlocale(LC_TIME, "");
		(void)setlocale(LC_CTYPE, "");
	}

	(void)setcat("uxcore.abi");
	(void)setlabel("UX:date");

	/*  Initialize variables  */
	(void)time(&clock_val);

	if (argc > 1) {
		if (strcmp(argv[1], "-u") == 0) {
			argc--; argv++;
			uflag++;
		}
		else if (strcmp(argv[1], "-n") == 0) {
			argc--; argv++;
			nflag++;
		}
		else if (strcmp(argv[1], "-a") == 0) {
			if (argc < 3){
				usage(1);
				exit(1);
			}
			get_adj(argv[2], &tv);

			retval = adjtime(&tv, 0);
			if( retval < 0 ) {
				pfmt(stderr, 
				  MM_ERROR, ":142:Cannot adjust date: %s\n",
					strerror(errno));
				exit(1);
			}
			exit(0);
		}		
	}
	if (argc > 1) {	
		if (*argv[1] == '+') {
			fmt = &argv[1][1];
			vflag++;
		}
                else if (strcmp(argv[1], "--") != 0) {
                        if (setdate(localtime(&clock_val), argv[1]))
                                exit(1);
                }
                else if (*argv[2] == '+') {
                        fmt = &argv[2][1];
                        vflag++;
                }
                else if (setdate(localtime(&clock_val), argv[2]))
                                exit(1);
	}

	if (uflag && vflag) {
		tp = gmtime(&clock_val);
		strftime(buf, sizeof(buf), fmt, tp);	
	} else if (uflag && !vflag) {
		tp = gmtime(&clock_val);
		strftime(buf, sizeof(buf),
		   (const char *)gettxt(":143", "%a %b %e %H:%M:%S GMT %Y"),
			tp);
		} else {	/* !uflag */
			tp = localtime(&clock_val);
			strftime(buf, sizeof(buf), fmt, tp);
		}

	puts(buf);

	exit(0);
}

/*
 * Procedure:     setdate
 */
setdate(current_date, date)
struct tm	*current_date;
char		*date;
{
	register int i;			/* scratch variable */
	int mm, dd=0, hh, min, yy, wf=0;
	int minidx = 6;
	static int tdate[4];
	
	static char badconv[] = ":144:Bad conversion\n";

#ifdef sgi
        char *sp, *p;
        int sec = 0;

        if (0 != (sp = strchr(date,'.'))) {
                sec = strtol(sp+1,&p,10);
                if (sec >= 0 && sec < 60 && *p == '\0')
                        *sp = '\0';
                else
                        date[0] = '\0';         /* To force error msg */
        }
#endif 


	/*  Parse date string  */
	switch(strlen(date)) {
	case 12:
		yy = atoi(&date[8]);
                if (yy < 1970 || yy > 2037) {
                        (void) pfmt(stderr, MM_ERROR, badconv);
                        usage(0);
                        return(1);
                }
		date[8] = '\0';
		break;
	case 10:
		i = atoi(&date[8]);
#ifdef old /* Bug#538632 */
		if(i >= 69 && i <= 99)		/* Same as touch */
			yy = 1900 + i;
		else	yy = 2000 + i;
#endif /* old */
                if (i >= 0 && i < 38)
                        yy = 2000 + i;
                else if (i > 69 && i <= 99)
                        yy = 1900 + i;
                else {
                        (void) pfmt(stderr, MM_ERROR, badconv);
                        usage(0);
                        return(1);
                }
		date[8] = '\0';
		break;
	case 8:
		yy = 1900 + current_date->tm_year;
		break;
	case 4: 
		yy = 1900 + current_date->tm_year;
		mm = current_date->tm_mon + 1; 	/* tm_mon goes from 1 to 11*/
		dd = current_date->tm_mday;
		minidx = 2;
		break;	
	default:
		(void) pfmt(stderr, MM_ERROR, badconv);
		usage(0);
		return(1);
	}
	tdate[0] = (int)date[minidx-2];
	tdate[1] = (int)date[minidx-1];
	tdate[2] = (int)date[minidx];
	tdate[3] = (int)date[minidx+1];
	for (i = 0; i < 4; i++) {
		if(!isdigit((int)tdate[i])) {
			(void) pfmt(stderr, MM_ERROR, badconv);
			usage(0);
			return(1);
		}
	}
	min = atoi(&date[minidx]);
	date[minidx] = '\0';
	hh = atoi(&date[minidx-2]);
	date[minidx-2] = '\0';
	if (!dd) { 
		/* if dd is 0 (not between 1 and 31), then 
		 * read the value supplied by the user.
		 */
		dd = atoi(&date[2]);
		date[2] = '\0';
		mm = atoi(&date[0]);
	}
	if(hh == 24)
		hh = 0, dd++;

	/*  Validate date elements  */
	if(!((yy<2100) && (mm >= 1 && mm <= 12) && (dd >= 1 && dd <= 31) &&
		(hh >= 0 && hh <= 23) && (min >= 0 && min <= 59))) {
		(void) pfmt(stderr, MM_ERROR, badconv);
		usage(0);
		return(1);
	}

	/*  Build date and time number  */
	for(clock_val = 0, i = 1970; i < yy; i++)
		clock_val += year_size(i);
	/*  Adjust for leap year  */
	if (year_size(yy) == 366 && mm >= 3)
		clock_val += 1;
	/*  Adjust for different month lengths  */
	while(--mm)
		clock_val += (time_t)month_size[mm - 1];
	/*  Load up the rest  */
	clock_val += (time_t)(dd - 1);
	clock_val *= 24;
	clock_val += (time_t)hh;
	clock_val *= 60;
	clock_val += (time_t)min;
	clock_val *= 60;
#ifdef sgi
        clock_val += sec;
#endif

	if (!uflag) {
		/* convert to GMT assuming standard time */
		/* correction is made in localtime(3C) */

		clock_val += (time_t)timezone;


		/* correct if daylight savings time in effect */

		if (localtime(&clock_val)->tm_isdst)
			clock_val = clock_val - (time_t)(timezone - altzone); 

	}

	(void) time(&wtmp[0].ut_time);


#ifdef sgi
	if ((nflag || !settime(clock_val)) && stime(&clock_val) < 0) {
#else
	if(stime(&clock_val) < 0) {
#endif
		(void) pfmt(stderr, MM_ERROR|MM_NOGET, "%s\n", strerror(errno));
		return(1);
	}
	(void) time(&wtmp[1].ut_time);

	/* Update utmp file - if update fails, just continue */
	pututline(&wtmp[0]);
	pututline(&wtmp[1]);
	
	wf = open(WTMP_FILE, O_WRONLY | O_APPEND);
	if ( wf >= 0) {
		(void)write(wf, (char *)wtmp, sizeof(wtmp));
		close(wf);
	} 
#ifdef sgi
	/*
	 * convert the wtmp entries into utmpx format in wtmpx[0..1]
	 * and append to /var/adm/wtmpx to keep wtmp and wtmpx in sync.
	 */
	getutmpx(&wtmp[0], &wtmpx[0]);
	getutmpx(&wtmp[1], &wtmpx[1]);
	wf = open(WTMPX_FILE, O_WRONLY|O_APPEND);
	if ( wf >= 0) {
		(void)write(wf, (char *)wtmpx, sizeof(wtmpx));
		close(wf);
	} 
#endif

	return(0);
}

/*
 * Procedure:     get_adj
 */
get_adj(cp, tp)
        char *cp;
        struct timeval *tp;
{
        register int mult;
        int sign;

        tp->tv_sec = tp->tv_usec = 0;
        if (*cp == '-') {
                sign = -1;
                cp++;
        } else {
                sign = 1;
        }
        while (*cp >= '0' && *cp <= '9') {
                tp->tv_sec *= 10;
                tp->tv_sec += *cp++ - '0';
        }
        if (*cp == '.') {
                cp++;
                mult = 100000;
                while (*cp >= '0' && *cp <= '9') {
                        tp->tv_usec += (*cp++ - '0') * mult;
                        mult /= 10;
                }
        }
        if (*cp){
        	usage(1);
        	exit(1);
        }

        tp->tv_sec *= sign;
        tp->tv_usec *= sign;
}

/*
 * Procedure:     usage
 */
usage(complain)
int complain;
{
 	if (complain)
		pfmt(stderr, MM_ERROR, ":8:Incorrect usage\n");
	pfmt(stderr, MM_ACTION,
		"uxsgicore:88:Usage:\n\tdate [-u] [+format]\n\tdate [-u | -n] [[mmdd] HHMM | mmddHHMM[[cc]yy]][.ss]\n\tdate [-a [-]sss.fff]\n");
}

#ifdef sgi
#include <sys/param.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <protocols/timed.h>


/*
 * Set the date in the machines controlled by timedaemons
 * by communicating the new date to the local timedaemon.
 * If the timedaemon is in the master state, it performs the
 * correction on all slaves.  If it is in the slave state, it
 * notifies the master that a correction is needed.
 * Returns 1 on success, 0 on failure.
 */
settime(tv)
time_t tv;
{
	int s, found, length, port, err;
	fd_set ready;
	char hostname[MAXHOSTNAMELEN+1];
	struct timeval tout;
	struct servent *sp;
	struct tsp msg;
	struct sockaddr_in sin, dest, from;

	sp = getservbyname("timed", "udp");
	if (sp == 0) {
		(void) pfmt(stderr, MM_ACTION, "uxsgicore:89:udp/timed: unknown service\n");
		return (0);
	}
	dest.sin_port = sp->s_port;
	dest.sin_family = AF_INET;
	dest.sin_addr.s_addr = htonl((u_long)INADDR_ANY);
	s = socket(AF_INET, SOCK_DGRAM, 0);
	if (s < 0) {
		if (errno != EPROTONOSUPPORT)
			(void) pfmt(stderr, MM_ACTION, "uxsgicore:3:socket\n");
		goto bad;
	}
	bzero((char *)&sin, sizeof (sin));
	sin.sin_family = AF_INET;
	for (port = IPPORT_RESERVED - 1; port > IPPORT_RESERVED / 2; port--) {
		sin.sin_port = htons((u_short)port);
		if (bind(s, (struct sockaddr *)&sin, sizeof (sin)) >= 0)
			break;
		if (errno != EADDRINUSE) {
			if (errno != EADDRNOTAVAIL)
				(void) pfmt(stderr, MM_ACTION, "uxsgicore:90:bind\n");
			goto bad;
		}
	}
	if (port == IPPORT_RESERVED / 2) {
		(void) pfmt(stderr, MM_ACTION, "uxsgicore:91:all ports in use\n");
		goto bad;
	}
	msg.tsp_type = TSP_SETDATE;
	msg.tsp_vers = TSPVERSION;
	(void) gethostname(hostname, sizeof (hostname));
	(void) strncpy(msg.tsp_name, hostname, sizeof (hostname));
	msg.tsp_seq = 0;
	msg.tsp_time.tv_sec = htonl(tv);
	msg.tsp_time.tv_usec = 0;
	if (connect(s, &dest, sizeof(struct sockaddr_in)) < 0) {
		(void) pfmt(stderr, MM_ACTION, "uxsgicore:92:connect\n");
		goto bad;
	}
	if (send(s, (char *)&msg, sizeof (struct tsp), 0) < 0) {
		if (errno == ECONNREFUSED) {
			(void)close(s);	/* no daemon */
			return (0);
		}
		(void) pfmt(stderr, MM_ACTION, "uxsgicore:93:send\n");
		goto bad;
	}

	FD_ZERO(&ready);
	FD_SET(s, &ready);
	for (;;) {
		tout.tv_sec = 5;
		tout.tv_usec = 0;
		found = select(FD_SETSIZE, &ready,
			       (fd_set *)0, (fd_set *)0, &tout);

		length = sizeof(err);
		if (getsockopt(s, SOL_SOCKET, SO_ERROR,
			       (char *)&err, &length) == 0
		    && err) {
			errno = err;
			if (errno == ECONNREFUSED)
				return (0);
			(void) pfmt(stderr, MM_ACTION, "uxsgicore:94:send (delayed error)");
			break;
		}

		if (found <= 0)
			break;

		length = sizeof (struct sockaddr_in);
		if (recvfrom(s, (char *)&msg, sizeof (struct tsp), 0,
			     &from, &length) < 0) {
			if (errno != ECONNREFUSED)
				(void) pfmt(stderr, MM_ACTION, "uxsgicore:95:recvform");
			break;
		}

		if (msg.tsp_type == TSP_ACK
		    || msg.tsp_type == TSP_DATEACK) {
			(void)close(s);
			return (1);
		}
	}
bad:
	(void) pfmt(stderr, MM_ACTION,
		"uxsgicore:96: Can't reach time daemon, time set locally.\n");
	(void)close(s);
	return (0);
}
#endif
